### PR TITLE
Introduce filter table [DPP-714]

### PR DESCRIPTION
**Pull request by nmarton-da**
* Adding new table, indexes
* Adding data migration (PostgreSQL only)
* Removing unnecessary indexes
* Adding DbDto, Schema integration for the filter table
* Adding generation of CreateFilter Dto-s to UpdateToDbDto
* Adding sequential ID generation to sandbox-classic and parallel-ingestion
* Adding StorageBackend support for pruning, reset
* Adding EventStorageBackend support for two phase retrival queries
* Extending/adapting unit tests
* Adding PostgreSQL VACUUM ANALYZE Flyway migration script as well (since this is the last persistence change in the ACS feature track)

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V116__add_filter_table.sql [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+INSERT INTO participant_events_create_filter(event_sequential_id, template_id, party_id)
+SELECT event_sequential_id, template_id, unnest(flat_event_witnesses)
+FROM participant_events_create
+ORDER BY event_sequential_id;
+
+SELECT 'Add Filter Table: Creating indexes...';
+
+CREATE INDEX idx_participant_events_create_filter_party_template_seq_id_idx ON participant_events_create_filter(party_id, template_id, event_sequential_id);
+CREATE INDEX idx_participant_events_create_filter_party_seq_id_idx ON participant_events_create_filter(party_id, event_sequential_id);
+CREATE INDEX idx_participant_events_create_seq_id_idx ON participant_events_create_filter(event_sequential_id);
```
Comment (nmarton-da) [inline ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V116__add_filter_table.sql:21]: @meiersi-da it's a pity: but we need this in order to efficiently cleanup the over-spill entries for parallel ingestion. Do you know maybne a way to somehow "reuse" the above indexes for this purpose? :pray:

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V117__vacuum_analyze.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+--  Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+--  SPDX-License-Identifier: Apache-2.0
+
+VACUUM ANALYZE;
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V117__vacuum_analyze.sql:4]: Is this the last step in the chain of migrations that implement the new ACS queries?

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala [hunk 3]
```diff
@@ -30,6 +30,7 @@ private[backend] trait StorageBackendTestsInitializeIngestion
       dtoPackageEntry(offset(3)),
       // 4: transaction with create node
       dtoCreate(offset(4), 1L, "#4"),
+      DbDto.CreateFilter(1L, someTemplateId.toString, someParty.toString),
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala:33]:
```suggestion
      dtoCreateFilter(1L, someParty),
```
we could add a helper like for all other DTOs

File: ledger/participant-integration-api/src/test/suite/scala/platform/indexer/parallel/ParallelIndexerSubscriptionSpec.scala [hunk 4]
```diff
@@ -195,6 +195,8 @@ class ParallelIndexerSubscriptionSpec extends AnyFlatSpec with Matchers {
           someEventDivulgence,
           someParty,
           someEventCreated,
+          DbDto.CreateFilter(0L, "", ""),
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/indexer/parallel/ParallelIndexerSubscriptionSpec.scala:198]:
```suggestion
          someCreateFilter,
```
for consistency

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala [hunk 5]
```diff
@@ -47,6 +48,7 @@ private[backend] trait StorageBackendTestsInitializeIngestion
       dtoPackageEntry(offset(8)),
       // 9: transaction with create node
       dtoCreate(offset(9), 4L, "#9"),
+      DbDto.CreateFilter(4L, someTemplateId.toString, someParty.toString),
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala:51]: There are many places where we ingest a create event without a corresponding filter event. The same already happens with completion events (not every `dtoCreate` is followed by a `dtoCompletion`) - the missing DTOs are not relevant for the respective test.
I'm still slightly worried about the fact that we have an increasing number of DTO sequences that cannot be produced by a real indexer.

**Review by rautenrieth-da — APPROVED** [review]

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V117__vacuum_analyze.sql (hunk 2 above)
Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V117__vacuum_analyze.sql:4]: exactly

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala (hunk 3 above)
Comment (nmarton-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala:33]: sure why not

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala (hunk 5 above)
Comment (nmarton-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/backend/StorageBackendTestsInitializeIngestion.scala:51]: The purpose of these tests is not to test the real indexer, rather some targeted functionality. These are unit tests.

File: ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V116__add_filter_table.sql (hunk 1 above)
Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/resources/db/migration/postgres-appendonly/V116__add_filter_table.sql:21]: I was always expecting to required the index on `event_sequential_id`. It's required for pruning anyways isn't it?

**Pull request merged**